lib/dry/schema/processor_steps.rb

```diff
@@ -120,18 +122,20 @@ def merge(other)
       # @api private
       def merge_callbacks(left, right)
         left.merge(right) do |_key, oldval, newval|
-          oldval + newval
+          (oldval + newval).sort_by(&:path)
         end
       end
 
       # @api private
       def import_callbacks(path, other)
         other.before_steps.each do |name, steps|
-          (before_steps[name] ||= []).concat(steps.map { |step| step.scoped(path) })
+          before_steps[name] ||= []
+          before_steps[name].concat(steps.map { |step| step.scoped(path) }).sort_by!(&:path)
         end
 
         other.after_steps.each do |name, steps|
-          (after_steps[name] ||= []).concat(steps.map { |step| step.scoped(path) })
+          after_steps[name] ||= []
+          after_steps[name].concat(steps.map { |step| step.scoped(path) }).sort_by!(&:path)
         end
       end
     end
```

Comment:
Probably we could sort callbacks just once instead of on each import/merge, but we're doing it only on schema creation, thus probably doesn't warrant additional complexity.